homeassistant/components/ps4/__init__.py

```diff
@@ -0,0 +1,215 @@
+"""
+Support for PlayStation 4 consoles.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/ps4/
+"""
+
+from collections import OrderedDict
+import logging
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_CODE, CONF_HOST, CONF_IP_ADDRESS,
+    CONF_NAME, CONF_REGION, CONF_TOKEN)
+from homeassistant.config_entries import SOURCE_USER
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'ps4'
+
+REGIONS = ('R1', 'R2', 'R3', 'R4', 'R5')
+DEFAULT_NAME = "PlayStation 4"
+DEFAULT_REGION = "R1"
+
+
+REQUIREMENTS = ['pyps4-homeassistant==0.2.5']
+
+
+async def async_setup(hass, config):
+    """Set up the PS4 Config Flow."""
+    hass.async_create_task(hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={'source': SOURCE_USER},
+        data=None))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up PS4 from a config entry."""
+    if hass.config_entries.async_entries(DOMAIN):
+        hass.data[DOMAIN] = config_entry
+        hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+            config_entry, 'media_player'))
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a PS4 config entry."""
+    await hass.config_entries.async_forward_entry_unload(
+        entry, 'media_player')
+
+    return True
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class PlayStation4FlowHandler(config_entries.ConfigFlow):
+    """Handle a PlayStation 4 config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.creds = None
+        self.name = None
+        self.host = None
+        self.device_id = None
+        self.region = None
+        self.pin = None
+        self.info = {}
+
+    async def _async_has_devices(self):
+        """Return if there are devices that can be discovered."""
+        from pyps4_homeassistant import ddp
+
+        _LOGGER.debug("Searching for PS4 Devices")
+        devices = await self.hass.async_add_executor_job(ddp.search)
+        for device in devices:
+            _LOGGER.debug("Found PS4 at: %s", device['host-ip'])
+        return devices
+
+    async def _async_link(self):
+        """Perform pairing with PS4."""
+        import pyps4_homeassistant as pyps4
+        ps4 = pyps4.Ps4(self.host, self.creds)
+        try:
+            await self.hass.async_add_executor_job(
+                ps4.login, self.pin)
+        except pyps4.errors.LoginFailed:
+            return False
+        except pyps4.errors.NotReady:
+            return 'not_ready'
```

Comment:
I just returned two boolean vars. Hope thats okay.

Reply:
It's ok. I suggested a namedtuple to allow adding more return values later without breaking code that relies on the method. But since this is a private method, I don't think we have to worry about that.